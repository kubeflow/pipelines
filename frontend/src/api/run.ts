// Copyright 2018 Google LLC
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//      http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

export class RunMetadata {
  public id = '';
  public name = '';
  public namespace = '';
  public created_at = '';
  public scheduled_at = '';
  public status = '';

  public static buildFromObject(run: any): RunMetadata {
    const newRunMetadata = new RunMetadata();
    newRunMetadata.id = run.id;
    newRunMetadata.name = run.name;
    newRunMetadata.namespace = run.namespace;
    newRunMetadata.created_at = run.created_at;
    newRunMetadata.scheduled_at = run.scheduled_at;
    newRunMetadata.status = run.status;
    return newRunMetadata;
  }
}

export class Run {
  public run = RunMetadata.buildFromObject({});
  public workflow = '';

  public static buildFromObject(oldRun: any): Run {
    const newRun = new Run();
    newRun.run = RunMetadata.buildFromObject(oldRun.run);
    newRun.workflow = oldRun.workflow;
    return newRun;
  }
}
