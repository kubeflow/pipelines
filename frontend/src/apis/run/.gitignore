wwwroot/*.js
node_modules
typings

# Ignore git_push script
git_push.sh
