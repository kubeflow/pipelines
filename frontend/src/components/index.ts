import { MyTopBar } from './top-bar';

// add custom elements here
const elements = {
  MyTopBar,
};

function toKebabCase(name: string) {
  return name.replace(/([a-z])([A-Z])/g, '$1-$2').toLowerCase();
}

// register all components as kebab case
Object.keys(elements)
  .forEach(key => {
    customElements.define(toKebabCase(key), elements[key])
  });
