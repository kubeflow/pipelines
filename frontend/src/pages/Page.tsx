/*
 * Copyright 2018 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the 'License');
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an 'AS IS' BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as React from 'react';
import { RouteComponentProps } from 'react-router';
import { ToolbarProps } from '../components/Toolbar';
import { BannerProps } from '../components/Banner';
import { SnackbarProps } from '@material-ui/core/Snackbar';
import { DialogProps } from '../components/Router';
import { errorToMessage } from '../lib/Utils';

export interface PageProps extends RouteComponentProps {
  toolbarProps: ToolbarProps;
  updateBanner: (bannerProps: BannerProps) => void;
  updateDialog: (dialogProps: DialogProps) => void;
  updateSnackbar: (snackbarProps: SnackbarProps) => void;
  updateToolbar: (toolbarProps: Partial<ToolbarProps>) => void;
}

export abstract class Page<P, S> extends React.Component<P & PageProps, S> {
  protected _isMounted = true;

  constructor(props: any) {
    super(props);
    this.props.updateToolbar(this.getInitialToolbarState());
  }

  public abstract render(): JSX.Element;

  public abstract getInitialToolbarState(): ToolbarProps;

  public abstract refresh(): Promise<void>;

  public componentWillUnmount(): void {
    this.clearBanner();
    this._isMounted = false;
  }

  public componentDidMount(): void {
    this.clearBanner();
  }

  public clearBanner(): void {
    this.props.updateBanner({});
  }

  public async showPageError(message: string, error?: Error, mode?: 'error' | 'warning'): Promise<void> {
    const errorMessage = await errorToMessage(error);
    this.props.updateBanner({
      additionalInfo: errorMessage ? errorMessage : undefined,
      message: message + (errorMessage ? ' Click Details for more information.' : ''),
      mode: mode || 'error',
      refresh: this.refresh.bind(this),
    });
  }

  public showErrorDialog(title: string, content: string): void {
    this.props.updateDialog({
      buttons: [{ text: 'Dismiss' }],
      content,
      title,
    });
  }

  protected setStateSafe(newState: Partial<S>, cb?: () => void): void {
    if (this._isMounted) {
      this.setState(newState as any, cb);
    }
  }
}
