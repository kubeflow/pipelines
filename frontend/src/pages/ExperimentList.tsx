/*
 * Copyright 2018 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as React from 'react';
import Buttons, { ButtonKeys } from '../lib/Buttons';
import CustomTable, { Column, Row, ExpandState, CustomRendererProps } from '../components/CustomTable';
import RunList from './RunList';
import produce from 'immer';
import { ApiFilter, PredicateOp } from '../apis/filter';
import { ApiListExperimentsResponse, ApiExperiment } from '../apis/experiment';
import { ApiResourceType, ApiRun, RunStorageState } from '../apis/run';
import { Apis, ExperimentSortKeys, ListRequest, RunSortKeys } from '../lib/Apis';
import { Link } from 'react-router-dom';
import { NodePhase } from '../lib/StatusUtils';
import { Page } from './Page';
import { RoutePage, RouteParams } from '../components/Router';
import { ToolbarProps } from '../components/Toolbar';
import { classes } from 'typestyle';
import { commonCss, padding } from '../Css';
import { logger } from '../lib/Utils';
import { statusToIcon } from './Status';

interface DisplayExperiment extends ApiExperiment {
  last5Runs?: ApiRun[];
  error?: string;
  expandState?: ExpandState;
}

interface ExperimentListState {
  displayExperiments: DisplayExperiment[];
  selectedIds: string[];
  selectedTab: number;
}

class ExperimentList extends Page<{}, ExperimentListState> {
  private _tableRef = React.createRef<CustomTable>();

  constructor(props: any) {
    super(props);

    this.state = {
      displayExperiments: [],
      selectedIds: [],
      selectedTab: 0,
    };
  }

  public getInitialToolbarState(): ToolbarProps {
    const buttons = new Buttons(this.props, this.refresh.bind(this));
    return {
      actions: buttons
        .newRun()
        .newExperiment()
        .compareRuns(() => this.state.selectedIds)
        .cloneRun(() => this.state.selectedIds, false)
        .archive(
          () => this.state.selectedIds,
          false,
          ids => this._selectionChanged(ids),
        )
        .refresh(this.refresh.bind(this))
        .getToolbarActionMap(),
      breadcrumbs: [],
      pageTitle: 'Experiments',
    };
  }

  public render(): JSX.Element {
    const columns: Column[] = [{
      customRenderer: this._nameCustomRenderer,
      flex: 1,
      label: 'Experiment name',
      sortKey: ExperimentSortKeys.NAME,
    }, {
      flex: 2,
      label: 'Description',
    }, {
      customRenderer: this._last5RunsCustomRenderer,
      flex: 1,
      label: 'Last 5 runs',
    }];

    const rows: Row[] = this.state.displayExperiments.map((exp) => {
      return {
        error: exp.error,
        expandState: exp.expandState,
        id: exp.id!,
        otherFields: [
          exp.name!,
          exp.description!,
          exp.expandState === ExpandState.EXPANDED ? [] : exp.last5Runs,
        ]
      };
    });

    return (
      <div className={classes(commonCss.page, padding(20, 'lr'))}>
        <CustomTable columns={columns} rows={rows} ref={this._tableRef}
          disableSelection={true} initialSortColumn={ExperimentSortKeys.CREATED_AT}
          reload={this._reload.bind(this)} toggleExpansion={this._toggleRowExpand.bind(this)}
          getExpandComponent={this._getExpandedExperimentComponent.bind(this)}
          filterLabel='Filter experiments'
          emptyMessage='No experiments found. Click "Create experiment" to start.' />
      </div>
    );
  }

  public async refresh(): Promise<void> {
    if (this._tableRef.current) {
      this.clearBanner();
      await this._tableRef.current.reload();
    }
  }

  public _nameCustomRenderer: React.FC<CustomRendererProps<string>> = (props: CustomRendererProps<string>) => {
    return <Link className={commonCss.link} onClick={(e) => e.stopPropagation()}
      to={RoutePage.EXPERIMENT_DETAILS.replace(':' + RouteParams.experimentId, props.id)}>{props.value}</Link>;
  }

  public _last5RunsCustomRenderer: React.FC<CustomRendererProps<ApiRun[]>> = (props: CustomRendererProps<ApiRun[]>) => {
    return <div className={commonCss.flex}>
      {(props.value || []).map((run, i) => (
        <span key={i} style={{ margin: '0 1px' }}>
          {statusToIcon(run.status as NodePhase || NodePhase.UNKNOWN, run.created_at)}
        </span>
      ))}
    </div>;
  }

  private async _reload(request: ListRequest): Promise<string> {
    // Fetch the list of experiments
    let response: ApiListExperimentsResponse;
    let displayExperiments: DisplayExperiment[];
    try {
      response = await Apis.experimentServiceApi.listExperiment(
        request.pageToken, request.pageSize, request.sortBy, request.filter);
      displayExperiments = response.experiments || [];
      displayExperiments.forEach((exp) => exp.expandState = ExpandState.COLLAPSED);
    } catch (err) {
      await this.showPageError('Error: failed to retrieve list of experiments.', err);
      // No point in continuing if we couldn't retrieve any experiments.
      return '';
    }

    // Fetch and set last 5 runs' statuses for each experiment
    await Promise.all(displayExperiments.map(async experiment => {
      // TODO: should we aggregate errors here? What if they fail for different reasons?
      try {
        const listRunsResponse = await Apis.runServiceApi.listRuns(
          undefined /* pageToken */,
          5 /* pageSize */,
          RunSortKeys.CREATED_AT + ' desc',
          ApiResourceType.EXPERIMENT.toString(),
          experiment.id,
          encodeURIComponent(JSON.stringify({
            predicates: [{
              key: 'storage_state',
              op: PredicateOp.NOTEQUALS,
              string_value: RunStorageState.ARCHIVED.toString(),
            }]
          } as ApiFilter)),
        );
        experiment.last5Runs = listRunsResponse.runs || [];
      } catch (err) {
        experiment.error = 'Failed to load the last 5 runs of this experiment';
        logger.error(
          `Error: failed to retrieve run statuses for experiment: ${experiment.name}.`,
          err);
      }
    }));

    this.setState({ displayExperiments });
    return response.next_page_token || '';
  }

  private _selectionChanged(selectedIds: string[]): void {
    const actions = this.props.toolbarProps.actions;
    actions[ButtonKeys.COMPARE].disabled = selectedIds.length <= 1 || selectedIds.length > 10;
    actions[ButtonKeys.CLONE_RUN].disabled = selectedIds.length !== 1;
    actions[ButtonKeys.ARCHIVE].disabled = !selectedIds.length;
    this.props.updateToolbar({ actions });
    this.setState({ selectedIds });
  }

  private _toggleRowExpand(rowIndex: number): void {
    const displayExperiments = produce(this.state.displayExperiments, draft => {
      draft[rowIndex].expandState =
        draft[rowIndex].expandState === ExpandState.COLLAPSED ?
          ExpandState.EXPANDED :
          ExpandState.COLLAPSED;
    });

    this.setState({ displayExperiments });
  }

  private _getExpandedExperimentComponent(experimentIndex: number): JSX.Element {
    const experiment = this.state.displayExperiments[experimentIndex];
    return <RunList hideExperimentColumn={true} experimentIdMask={experiment.id} onError={() => null} {...this.props}
      disablePaging={true} selectedIds={this.state.selectedIds} noFilterBox={true}
      storageState={RunStorageState.AVAILABLE}
      onSelectionChange={this._selectionChanged.bind(this)} disableSorting={true} />;
  }

}

export default ExperimentList;
