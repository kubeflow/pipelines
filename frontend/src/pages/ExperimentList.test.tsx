/*
 * Copyright 2018 Google LLC
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * https://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import * as React from 'react';
import * as Utils from '../lib/Utils';
import ExperimentList from './ExperimentList';
import TestUtils from '../TestUtils';
import { ApiFilter, PredicateOp } from '../apis/filter';
import { ApiResourceType, RunStorageState } from '../apis/run';
import { Apis } from '../lib/Apis';
import { ExpandState } from '../components/CustomTable';
import { NodePhase } from './Status';
import { PageProps } from './Page';
import { ReactWrapper, ShallowWrapper, shallow } from 'enzyme';
import { RoutePage, QUERY_PARAMS } from '../components/Router';
import { range } from 'lodash';

describe('ExperimentList', () => {
  let tree: ShallowWrapper | ReactWrapper;

  jest.spyOn(console, 'log').mockImplementation(() => null);

  const updateBannerSpy = jest.fn();
  const updateDialogSpy = jest.fn();
  const updateSnackbarSpy = jest.fn();
  const updateToolbarSpy = jest.fn();
  const historyPushSpy = jest.fn();
  const listExperimentsSpy = jest.spyOn(Apis.experimentServiceApi, 'listExperiment');
  const listRunsSpy = jest.spyOn(Apis.runServiceApi, 'listRuns');
  // We mock this because it uses toLocaleDateString, which causes mismatches between local and CI
  // test enviroments
  jest.spyOn(Utils, 'formatDateString').mockImplementation(() => '1/2/2019, 12:34:56 PM');

  function generateProps(): PageProps {
    return TestUtils.generatePageProps(ExperimentList, { pathname: RoutePage.EXPERIMENTS } as any,
      '' as any, historyPushSpy, updateBannerSpy, updateDialogSpy, updateToolbarSpy, updateSnackbarSpy);
  }

  async function mountWithNExperiments(n: number, nRuns: number): Promise<void> {
    listExperimentsSpy.mockImplementation(() => ({
      experiments: range(n).map(i => ({ id: 'test-experiment-id' + i, name: 'test experiment name' + i })),
    }));
    listRunsSpy.mockImplementation(() => ({
      runs: range(nRuns).map(i => ({ id: 'test-run-id' + i, name: 'test run name' + i })),
    }));
    tree = TestUtils.mountWithRouter(<ExperimentList {...generateProps()} />);
    await listExperimentsSpy;
    await listRunsSpy;
    await TestUtils.flushPromises();
    tree.update(); // Make sure the tree is updated before returning it
  }

  afterEach(() => {
    jest.resetAllMocks();
    jest.clearAllMocks();
    tree.unmount();
  });

  it('renders an empty list with empty state message', () => {
    tree = shallow(<ExperimentList {...generateProps()} />);
    expect(tree).toMatchSnapshot();
  });

  it('renders a list of one experiment', async () => {
    tree = shallow(<ExperimentList {...generateProps()} />);
    tree.setState({
      displayExperiments: [{
        description: 'test experiment description',
        expandState: ExpandState.COLLAPSED,
        name: 'test experiment name',
      }]
    });
    await listExperimentsSpy;
    await listRunsSpy;
    expect(tree).toMatchSnapshot();
  });

  it('renders a list of one experiment with no description', async () => {
    tree = shallow(<ExperimentList {...generateProps()} />);
    tree.setState({
      experiments: [{
        expandState: ExpandState.COLLAPSED,
        name: 'test experiment name',
      }]
    });
    await listExperimentsSpy;
    await listRunsSpy;
    expect(tree).toMatchSnapshot();
  });

  it('renders a list of one experiment with error', async () => {
    tree = shallow(<ExperimentList {...generateProps()} />);
    tree.setState({
      experiments: [{
        description: 'test experiment description',
        error: 'oops! could not load experiment',
        expandState: ExpandState.COLLAPSED,
        name: 'test experiment name',
      }]
    });
    await listExperimentsSpy;
    await listRunsSpy;
    expect(tree).toMatchSnapshot();
  });

  it('calls Apis to list experiments, sorted by creation time in descending order', async () => {
    await mountWithNExperiments(1, 1);
    expect(listExperimentsSpy).toHaveBeenLastCalledWith('', 10, 'created_at desc', '');
    expect(listRunsSpy).toHaveBeenLastCalledWith(
      undefined,
      5,
      'created_at desc',
      ApiResourceType.EXPERIMENT.toString(),
      'test-experiment-id0',
      encodeURIComponent(JSON.stringify({
        predicates: [{
          key: 'storage_state',
          op: PredicateOp.NOTEQUALS,
          string_value: RunStorageState.ARCHIVED.toString(),
        }],
      } as ApiFilter)),
    );
    expect(tree.state()).toHaveProperty('displayExperiments', [{
      expandState: ExpandState.COLLAPSED,
      id: 'test-experiment-id0',
      last5Runs: [{ id: 'test-run-id0', name: 'test run name0' }],
      name: 'test experiment name0',
    }]);
  });

  it('has a Refresh button, clicking it refreshes the experiment list', async () => {
    await mountWithNExperiments(1, 1);
    const instance = tree.instance() as ExperimentList;
    expect(listExperimentsSpy.mock.calls.length).toBe(1);
    const refreshBtn = instance.getInitialToolbarState().actions.find(b => b.title === 'Refresh');
    expect(refreshBtn).toBeDefined();
    await refreshBtn!.action();
    expect(listExperimentsSpy.mock.calls.length).toBe(2);
    expect(listExperimentsSpy).toHaveBeenLastCalledWith('', 10, 'created_at desc', '');
    expect(updateBannerSpy).toHaveBeenLastCalledWith({});
  });

  it('shows error banner when listing experiments fails', async () => {
    TestUtils.makeErrorResponseOnce(listExperimentsSpy, 'bad stuff happened');
    tree = TestUtils.mountWithRouter(<ExperimentList {...generateProps()} />);
    await listExperimentsSpy;
    await TestUtils.flushPromises();
    expect(updateBannerSpy).toHaveBeenLastCalledWith(expect.objectContaining({
      additionalInfo: 'bad stuff happened',
      message: 'Error: failed to retrieve list of experiments. Click Details for more information.',
      mode: 'error',
    }));
  });

  it('shows error next to experiment when listing its last 5 runs fails', async () => {
    // tslint:disable-next-line:no-console
    console.error = jest.spyOn(console, 'error').mockImplementation();

    listExperimentsSpy.mockImplementationOnce(() => ({ experiments: [{ name: 'exp1' }] }));
    TestUtils.makeErrorResponseOnce(listRunsSpy, 'bad stuff happened');
    tree = TestUtils.mountWithRouter(<ExperimentList {...generateProps()} />);
    await listExperimentsSpy;
    await TestUtils.flushPromises();
    expect(tree.state()).toHaveProperty('displayExperiments', [{
      error: 'Failed to load the last 5 runs of this experiment',
      expandState: 0,
      name: 'exp1',
    }]);
  });

  it('shows error banner when listing experiments fails after refresh', async () => {
    tree = TestUtils.mountWithRouter(<ExperimentList {...generateProps()} />);
    const instance = tree.instance() as ExperimentList;
    const refreshBtn = instance.getInitialToolbarState().actions.find(b => b.title === 'Refresh');
    expect(refreshBtn).toBeDefined();
    TestUtils.makeErrorResponseOnce(listExperimentsSpy, 'bad stuff happened');
    await refreshBtn!.action();
    expect(listExperimentsSpy.mock.calls.length).toBe(2);
    expect(listExperimentsSpy).toHaveBeenLastCalledWith('', 10, 'created_at desc', '');
    expect(updateBannerSpy).toHaveBeenLastCalledWith(expect.objectContaining({
      additionalInfo: 'bad stuff happened',
      message: 'Error: failed to retrieve list of experiments. Click Details for more information.',
      mode: 'error',
    }));
  });

  it('hides error banner when listing experiments fails then succeeds', async () => {
    TestUtils.makeErrorResponseOnce(listExperimentsSpy, 'bad stuff happened');
    tree = TestUtils.mountWithRouter(<ExperimentList {...generateProps()} />);
    const instance = tree.instance() as ExperimentList;
    await listExperimentsSpy;
    await TestUtils.flushPromises();
    expect(updateBannerSpy).toHaveBeenLastCalledWith(expect.objectContaining({
      additionalInfo: 'bad stuff happened',
      message: 'Error: failed to retrieve list of experiments. Click Details for more information.',
      mode: 'error',
    }));
    updateBannerSpy.mockReset();

    const refreshBtn = instance.getInitialToolbarState().actions.find(b => b.title === 'Refresh');
    listExperimentsSpy.mockImplementationOnce(() => ({ experiments: [{ name: 'experiment1' }] }));
    listRunsSpy.mockImplementationOnce(() => ({ runs: [{ name: 'run1' }] }));
    await refreshBtn!.action();
    expect(listExperimentsSpy.mock.calls.length).toBe(2);
    expect(updateBannerSpy).toHaveBeenLastCalledWith({});
  });

  it('can expand an experiment to see its runs', async () => {
    await mountWithNExperiments(1, 1);
    tree.find('.tableRow button').at(0).simulate('click');
    expect(tree.state()).toHaveProperty('displayExperiments', [{
      expandState: ExpandState.EXPANDED,
      id: 'test-experiment-id0',
      last5Runs: [{ id: 'test-run-id0', name: 'test run name0' }],
      name: 'test experiment name0',
    }]);
  });

  it('renders a list of runs for given experiment', async () => {
    tree = shallow(<ExperimentList {...generateProps()} />);
    tree.setState({ displayExperiments: [{ id: 'experiment1', last5Runs: [{ id: 'run1id' }, { id: 'run2id' }] }] });
    const runListTree = (tree.instance() as any)._getExpandedExperimentComponent(0);
    expect(runListTree.props.experimentIdMask).toEqual('experiment1');
  });

  it('navigates to new experiment page when Create experiment button is clicked', async () => {
    tree = TestUtils.mountWithRouter(<ExperimentList {...generateProps()} />);
    const createBtn = (tree.instance() as ExperimentList)
      .getInitialToolbarState().actions.find(b => b.title === 'Create experiment');
    await createBtn!.action();
    expect(historyPushSpy).toHaveBeenLastCalledWith(RoutePage.NEW_EXPERIMENT);
  });

  it('always has new experiment button enabled', async () => {
    await mountWithNExperiments(1, 1);
    const calls = updateToolbarSpy.mock.calls[0];
    expect(calls[0].actions.find((b: any) => b.title === 'Create experiment')).not.toHaveProperty('disabled');
  });

  it('enables clone button when one run is selected', async () => {
    await mountWithNExperiments(1, 1);
    (tree.instance() as any)._selectionChanged(['run1']);
    expect(updateToolbarSpy).toHaveBeenCalledTimes(2);
    expect(updateToolbarSpy.mock.calls[0][0].actions.find((b: any) => b.title === 'Clone run'))
      .toHaveProperty('disabled', true);
    expect(updateToolbarSpy.mock.calls[1][0].actions.find((b: any) => b.title === 'Clone run'))
      .toHaveProperty('disabled', false);
  });

  it('disables clone button when more than one run is selected', async () => {
    await mountWithNExperiments(1, 1);
    (tree.instance() as any)._selectionChanged(['run1', 'run2']);
    expect(updateToolbarSpy).toHaveBeenCalledTimes(2);
    expect(updateToolbarSpy.mock.calls[0][0].actions.find((b: any) => b.title === 'Clone run'))
      .toHaveProperty('disabled', true);
    expect(updateToolbarSpy.mock.calls[1][0].actions.find((b: any) => b.title === 'Clone run'))
      .toHaveProperty('disabled', true);
  });

  it('enables compare runs button only when more than one is selected', async () => {
    await mountWithNExperiments(1, 1);
    (tree.instance() as any)._selectionChanged(['run1']);
    (tree.instance() as any)._selectionChanged(['run1', 'run2']);
    (tree.instance() as any)._selectionChanged(['run1', 'run2', 'run3']);
    expect(updateToolbarSpy).toHaveBeenCalledTimes(4);
    expect(updateToolbarSpy.mock.calls[0][0].actions.find((b: any) => b.title === 'Compare runs'))
      .toHaveProperty('disabled', true);
    expect(updateToolbarSpy.mock.calls[1][0].actions.find((b: any) => b.title === 'Compare runs'))
      .toHaveProperty('disabled', true);
    expect(updateToolbarSpy.mock.calls[2][0].actions.find((b: any) => b.title === 'Compare runs'))
      .toHaveProperty('disabled', false);
    expect(updateToolbarSpy.mock.calls[3][0].actions.find((b: any) => b.title === 'Compare runs'))
      .toHaveProperty('disabled', false);
  });

  it('navigates to compare page with the selected run ids', async () => {
    await mountWithNExperiments(1, 1);
    (tree.instance() as any)._selectionChanged(['run1', 'run2', 'run3']);
    const compareBtn = (tree.instance() as ExperimentList)
      .getInitialToolbarState().actions.find(b => b.title === 'Compare runs');
    await compareBtn!.action();
    expect(historyPushSpy).toHaveBeenLastCalledWith(
      `${RoutePage.COMPARE}?${QUERY_PARAMS.runlist}=run1,run2,run3`);
  });

  it('navigates to new run page with the selected run id for cloning', async () => {
    await mountWithNExperiments(1, 1);
    (tree.instance() as any)._selectionChanged(['run1']);
    const cloneBtn = (tree.instance() as ExperimentList)
      .getInitialToolbarState().actions.find(b => b.title === 'Clone run');
    await cloneBtn!.action();
    expect(historyPushSpy).toHaveBeenLastCalledWith(
      `${RoutePage.NEW_RUN}?${QUERY_PARAMS.cloneFromRun}=run1`);
  });

  it('enables archive button when at least one run is selected', async () => {
    await mountWithNExperiments(1, 1);
    expect(TestUtils.getToolbarButton(updateToolbarSpy, 'Archive').disabled).toBeTruthy();
    (tree.instance() as any)._selectionChanged(['run1']);
    expect(TestUtils.getToolbarButton(updateToolbarSpy, 'Archive').disabled).toBeFalsy();
    (tree.instance() as any)._selectionChanged(['run1', 'run2']);
    expect(TestUtils.getToolbarButton(updateToolbarSpy, 'Archive').disabled).toBeFalsy();
    (tree.instance() as any)._selectionChanged([]);
    expect(TestUtils.getToolbarButton(updateToolbarSpy, 'Archive').disabled).toBeTruthy();
  });

  it('renders experiment names as links to their details pages', async () => {
    tree = TestUtils.mountWithRouter(<ExperimentList {...generateProps()} />);
    expect((tree.instance() as ExperimentList)._nameCustomRenderer({
      id: 'experiment-id',
      value: 'experiment name',
    })).toMatchSnapshot();
  });

  it('renders last 5 runs statuses', async () => {
    tree = TestUtils.mountWithRouter(<ExperimentList {...generateProps()} />);
    expect((tree.instance() as ExperimentList)._last5RunsCustomRenderer({
      id: 'experiment-id',
      value: [
        { status: NodePhase.SUCCEEDED },
        { status: NodePhase.PENDING },
        { status: NodePhase.FAILED },
        { status: NodePhase.UNKNOWN },
        { status: NodePhase.SUCCEEDED },
      ],
    })).toMatchSnapshot();
  });
});
