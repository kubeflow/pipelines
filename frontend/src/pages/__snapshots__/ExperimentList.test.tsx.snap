// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExperimentList renders a list of one experiment 1`] = `
<div
  className="page"
>
  <CustomTable
    columns={
      Array [
        Object {
          "customRenderer": [Function],
          "flex": 1,
          "label": "Experiment name",
          "sortKey": "name",
        },
        Object {
          "flex": 2,
          "label": "Description",
        },
        Object {
          "customRenderer": [Function],
          "flex": 1,
          "label": "Last 5 runs",
        },
      ]
    }
    disableSelection={true}
    emptyMessage="No experiments found. Click \\"Create experiment\\" to start."
    filterLabel="Filter experiments"
    getExpandComponent={[Function]}
    initialSortColumn="created_at"
    reload={[Function]}
    rows={
      Array [
        Object {
          "error": undefined,
          "expandState": 0,
          "id": undefined,
          "otherFields": Array [
            "test experiment name",
            "test experiment description",
            undefined,
          ],
        },
      ]
    }
    toggleExpansion={[Function]}
  />
</div>
`;

exports[`ExperimentList renders a list of one experiment with error 1`] = `
<div
  className="page"
>
  <CustomTable
    columns={
      Array [
        Object {
          "customRenderer": [Function],
          "flex": 1,
          "label": "Experiment name",
          "sortKey": "name",
        },
        Object {
          "flex": 2,
          "label": "Description",
        },
        Object {
          "customRenderer": [Function],
          "flex": 1,
          "label": "Last 5 runs",
        },
      ]
    }
    disableSelection={true}
    emptyMessage="No experiments found. Click \\"Create experiment\\" to start."
    filterLabel="Filter experiments"
    getExpandComponent={[Function]}
    initialSortColumn="created_at"
    reload={[Function]}
    rows={Array []}
    toggleExpansion={[Function]}
  />
</div>
`;

exports[`ExperimentList renders a list of one experiment with no description 1`] = `
<div
  className="page"
>
  <CustomTable
    columns={
      Array [
        Object {
          "customRenderer": [Function],
          "flex": 1,
          "label": "Experiment name",
          "sortKey": "name",
        },
        Object {
          "flex": 2,
          "label": "Description",
        },
        Object {
          "customRenderer": [Function],
          "flex": 1,
          "label": "Last 5 runs",
        },
      ]
    }
    disableSelection={true}
    emptyMessage="No experiments found. Click \\"Create experiment\\" to start."
    filterLabel="Filter experiments"
    getExpandComponent={[Function]}
    initialSortColumn="created_at"
    reload={[Function]}
    rows={Array []}
    toggleExpansion={[Function]}
  />
</div>
`;

exports[`ExperimentList renders an empty list with empty state message 1`] = `
<div
  className="page"
>
  <CustomTable
    columns={
      Array [
        Object {
          "customRenderer": [Function],
          "flex": 1,
          "label": "Experiment name",
          "sortKey": "name",
        },
        Object {
          "flex": 2,
          "label": "Description",
        },
        Object {
          "customRenderer": [Function],
          "flex": 1,
          "label": "Last 5 runs",
        },
      ]
    }
    disableSelection={true}
    emptyMessage="No experiments found. Click \\"Create experiment\\" to start."
    filterLabel="Filter experiments"
    getExpandComponent={[Function]}
    initialSortColumn="created_at"
    reload={[Function]}
    rows={Array []}
    toggleExpansion={[Function]}
  />
</div>
`;

exports[`ExperimentList renders experiment names as links to their details pages 1`] = `
<WithStyles(Tooltip)
  enterDelay={300}
  placement="top-start"
  title="experiment name"
>
  <Link
    className="link"
    onClick={[Function]}
    replace={false}
    to="/experiments/details/experiment-id"
  >
    experiment name
  </Link>
</WithStyles(Tooltip)>
`;

exports[`ExperimentList renders last 5 runs statuses 1`] = `
<div
  className="flex"
>
  <span
    style={
      Object {
        "margin": "0 1px",
      }
    }
  >
    <WithStyles(Tooltip)
      title={
        <div>
          <div>
            Executed successfully
          </div>
        </div>
      }
    >
      <span
        style={
          Object {
            "height": 18,
          }
        }
      >
        <pure(CheckCircleIcon)
          data-testid="node-status-sign"
          style={
            Object {
              "color": "#34a853",
              "height": 18,
              "width": 18,
            }
          }
        />
      </span>
    </WithStyles(Tooltip)>
  </span>
  <span
    style={
      Object {
        "margin": "0 1px",
      }
    }
  >
    <WithStyles(Tooltip)
      title={
        <div>
          <div>
            Pending execution
          </div>
        </div>
      }
    >
      <span
        style={
          Object {
            "height": 18,
          }
        }
      >
        <pure(ScheduleIcon)
          data-testid="node-status-sign"
          style={
            Object {
              "color": "#9aa0a6",
              "height": 18,
              "width": 18,
            }
          }
        />
      </span>
    </WithStyles(Tooltip)>
  </span>
  <span
    style={
      Object {
        "margin": "0 1px",
      }
    }
  >
    <WithStyles(Tooltip)
      title={
        <div>
          <div>
            Resource failed to execute
          </div>
        </div>
      }
    >
      <span
        style={
          Object {
            "height": 18,
          }
        }
      >
        <pure(ErrorIcon)
          data-testid="node-status-sign"
          style={
            Object {
              "color": "#d50000",
              "height": 18,
              "width": 18,
            }
          }
        />
      </span>
    </WithStyles(Tooltip)>
  </span>
  <span
    style={
      Object {
        "margin": "0 1px",
      }
    }
  >
    <WithStyles(Tooltip)
      title={
        <div>
          <div>
            Unknown status
          </div>
        </div>
      }
    >
      <span
        style={
          Object {
            "height": 18,
          }
        }
      >
        <pure(HelpIcon)
          data-testid="node-status-sign"
          style={
            Object {
              "color": "#5f6368",
              "height": 18,
              "width": 18,
            }
          }
        />
      </span>
    </WithStyles(Tooltip)>
  </span>
  <span
    style={
      Object {
        "margin": "0 1px",
      }
    }
  >
    <WithStyles(Tooltip)
      title={
        <div>
          <div>
            Executed successfully
          </div>
        </div>
      }
    >
      <span
        style={
          Object {
            "height": 18,
          }
        }
      >
        <pure(CheckCircleIcon)
          data-testid="node-status-sign"
          style={
            Object {
              "color": "#34a853",
              "height": 18,
              "width": 18,
            }
          }
        />
      </span>
    </WithStyles(Tooltip)>
  </span>
</div>
`;
